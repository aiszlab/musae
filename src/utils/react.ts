import { Partialable } from "@aiszlab/relax/types";
import { Children, isValidElement, type ReactNode } from "react";

interface ElementProps {
  children?: ReactNode;
}

/**
 * @description
 * get the string of component children
 */
export const toReactNodeText = (children: ReactNode): Partialable<string> => {
  const _children = Children.toArray(children);

  for (const child of _children) {
    if (typeof child === "string" || typeof child === "number") {
      return child.toString();
    }

    if (isValidElement<ElementProps>(child)) {
      const textInChild = toReactNodeText(child.props.children);

      if (textInChild) {
        return textInChild;
      }
    }
  }
};
