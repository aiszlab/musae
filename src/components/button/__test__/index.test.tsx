import { fireEvent, render, getByText } from "@testing-library/react";
import Button from "../button";
import React from "react";
import "@testing-library/jest-dom";

describe("`Avatar` Component", () => {
  test("snapshot", () => {
    const { asFragment } = render(<Button>button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("disabled", () => {
    const click = jest.fn();

    const { container: disabledButton } = render(
      <Button disabled onClick={click}>
        disabled
      </Button>
    );
    const { container: normalButton } = render(<Button onClick={click}>normal</Button>);

    fireEvent.click(getByText(disabledButton, "disabled"));
    expect(click).toHaveBeenCalledTimes(0);
    fireEvent.click(getByText(normalButton, "normal"));
    expect(click).toHaveBeenCalledTimes(1);
  });
});
