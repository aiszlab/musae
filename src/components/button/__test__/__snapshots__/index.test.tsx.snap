// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Button\` Component disabled variant render 1`] = `
<DocumentFragment>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.filled button__styles.rounded button__styles.disabled"
    disabled=""
    style="--backgroundColor: rgba(29,27,32,0.12); --color: rgba(29,27,32,0.38); --borderColor: initial;"
    type="button"
  >
    disabled filled
  </button>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.outlined button__styles.rounded button__styles.disabled"
    disabled=""
    style="--b6ve2z: rgba(103,80,164,0.08); --backgroundColor: transparent; --color: rgba(29,27,32,0.38); --borderColor: rgba(29,27,32,0.38);"
    type="button"
  >
    disabled outlined
  </button>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.text button__styles.rounded button__styles.disabled"
    disabled=""
    style="--b6ve2z: rgba(103,80,164,0.08); --backgroundColor: transparent; --color: rgba(29,27,32,0.38); --borderColor: initial;"
    type="button"
  >
    disabled text
  </button>
</DocumentFragment>
`;

exports[`\`Button\` Component variant render 1`] = `
<DocumentFragment>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.filled button__styles.rounded"
    style="--backgroundColor: #6750A4; --color: #FFF;"
    type="button"
  >
    filled
  </button>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.outlined button__styles.rounded"
    style="--borderColor: #6750A4; --color: #6750A4; --b6ve2z: rgba(103,80,164,0.08);"
    type="button"
  >
    outlined
  </button>
  <button
    class="musae-button button__styles.button button__styles.ripple theme__label.medium button__styles.medium button__styles.text button__styles.rounded"
    style="--color: #6750A4; --b6ve2z: rgba(103,80,164,0.08);"
    type="button"
  >
    text
  </button>
</DocumentFragment>
`;
